### PR TITLE
Writing of lineage and usage, through app fabric

### PR DESCRIPTION
**Pull request by anwar6953**
Similar approach as https://github.com/caskdata/cdap/pull/5910

Writing updates to usage and lineage (remotely) through app fabric.

Pending (which will come in a separate PR):
1. Scalability of this service/handler.
2. Not using deprecated `Id` classes in new code. This will come as a separate change, since the new interface shares methods with existing interface and implementation.

https://issues.cask.co/browse/CDAP-6250
http://builds.cask.co/browse/CDAP-DUT4255

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+          bind(RuntimeStore.class).to(RemoteRuntimeStore.class);
+
+          // For binding StreamWriter
+          install(createStreamFactoryModule());
+        }
+      }
+    );
+
+    return Modules.override(combined).with(new AbstractModule() {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java:120]: why do you need to unconditionally override a module that was just created in this same method? Perhaps include these bindings in the `Modules.combine` statement above itself?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/AbstractRemoteStoreHandler.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.lang.reflect.Type;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Implements common functionality for reading method arguments and deserializing them.
+ */
+public class AbstractRemoteStoreHandler extends AbstractHttpHandler {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/AbstractRemoteStoreHandler.java]: Since this is not specific to "Store", should make this more generic. `AbstractRemoteSystemOpsHandler`? `AbstractRemoteProgramOpsHandler`

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+
+import java.util.Iterator;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * The {@link co.cask.http.HttpHandler} for handling REST calls to LineageStore.
+ */
+@Path(AbstractRemoteStoreHandler.VERSION + "/execute")
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java]: why do we need the `/execute` path prefix? Perhaps something like `/remoteops` instead?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java [hunk 4]
```diff
@@ -45,31 +46,31 @@
...
   private final int numGroups;
-  private final UsageRegistry usageRegistry;
-  private final Id.Namespace namespace;
+  private final RuntimeUsageRegistry runtimeUsageRegistry;
   private final List<Id> owners;
   private ConsumerConfig consumerConfig;
   private Closeable consumer;
 
-  static <T> ConsumerSupplier<T> create(Id.Namespace namespace, List<Id> owners, UsageRegistry usageRegistry,
+  static <T> ConsumerSupplier<T> create(List<Id> owners,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java:54]: why is namespace not needed anymore?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteStoreClient.java [hunk 5]
```diff
@@ -45,6 +55,12 @@
  */
 public class RemoteStoreClient {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteStoreClient.java]: Perhaps rename to make this more generic, since it seems to be used by all 'system/CDAP platform' ops in programs?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/DatasetUsageKey.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ * the License.
+ */
+
+package co.cask.cdap.data2.registry;
+
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Objects;
+
+/**
+ *
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/DatasetUsageKey.java]: empty javadoc?

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java (hunk 1 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java:120]: LineageWriter and RuntimeUsageRegistry are not explicitly bound in the Modules.combine` methods above. So, this was the only way I could see overriding it.
If I override it in the configure method above, it complained that the bindings were already defined.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java (hunk 3 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java]: I went with 'execute', since its effectively executing the remote operation.
Both 'execute' and what you suggested seem good to me.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java (hunk 4 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/flow/ConsumerSupplier.java:54]: It wasn't being used from the class when I started this change, so I simply removed the parameter passed in.
Intelij complained it was unused, I think.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java]: I suggested a plural noun to indicate a 'contains' relationship with the individual operations because that's more RESTful than using the verb 'execute'

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java:120]: LineageWriter is bound in DataSetsModules. It seems wrong to have that binding defined in this same method only to override it in the next statement. Perhaps get this reviewed with @chtyim. I think there is a conscious effort to not use Modules.override, unless absolutely necessary.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/AbstractRemoteSystemOpsHandler.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    MethodArgument argument = arguments.next();
+    if (argument == null) {
+      return null;
+    }
+    JsonElement value = argument.getValue();
+    if (value == null) {
+      return null;
+    }
+    return GSON.<T>fromJson(value, Class.forName(argument.getType()));
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/AbstractRemoteSystemOpsHandler.java:64]: could using `Class.forName` cause a problem here? Should you load the class from the context classloader instead?

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  public RemoteLineageWriterHandler(LineageWriter lineageWriter) {
+    this.lineageWriter = lineageWriter;
+  }
+
+  @POST
+  @Path("/addDatasetAccess")
+  public void addDatasetAccess(HttpRequest request, HttpResponder responder) throws Exception {
+    Iterator<MethodArgument> arguments = parseArguments(request);
+
+    Id.Run run = deserializeNext(arguments);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java:50]: still think your follow-up PR for using the new entity classes could have a lot less code if you used the new entity classes here, and just used `toId` in the LineageWriter methods. But I'll leave that to you :-)

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteOpsClient.java [hunk 9]
```diff
@@ -36,20 +44,28 @@
...
 import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 import javax.annotation.Nullable;
 
 /**
- * Common HTTP client functionality for remote store implementations.
+ * Common HTTP client functionality for remote operation implementations.
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteOpsClient.java]: nit: ... for remote operations in programs?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteOpsClient.java [hunk 10]
```diff
@@ -73,6 +89,24 @@ private String resolve(String resource) {
                          addr.getHostName(), addr.getPort(), "/v1", resource);
   }
 
+
+  protected void executeRequest(String methodName, Object... arguments) {
+    doPost("execute/" + methodName, GSON.toJson(createArguments(arguments)));
+  }
+
+  private static List<MethodArgument> createArguments(Object... arguments) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteOpsClient.java:97]: nit: do we need this method? only seems to be called once. Also, if needed, does it need to be static?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/RuntimeUsageRegistry.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Store program -> dataset/stream usage information. Differs from UsageRegistry in that UsageRegistry does not have
+ * a remote implementation, usable from program runtime.
+ */
+public interface RuntimeUsageRegistry {
+
+  /**
+   * Registers usage of a stream by multiple ids.
+   *
+   * @param users    the users of the stream
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/RuntimeUsageRegistry.java]: nit: use single space, everywhere in this file

File: cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java [hunk 12]
```diff
@@ -127,6 +127,7 @@ private static Context createContext(Configuration conf) throws IOException {
       new LocationRuntimeModule().getDistributedModules(),
       new DiscoveryRuntimeModule().getDistributedModules(),
       new DataFabricModules().getDistributedModules(),
+      // here
```
Comment (bdmogal) [inline cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java]: did you mean to fix something here? else remove comment?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java [hunk 13]
```diff
@@ -126,31 +105,6 @@ protected void shutDown() throws Exception {
   }
 
   private static Injector createInjector(CConfiguration cConf, Configuration hConf) {
-    return Guice.createInjector(
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java:129]: there seem to be some differences between these bindings and the ones in `AbstractProgramTwillRunnable`. Ok to refactor them as `DistributedProgramRunnableModule`?

**Comment by bdmogal** [comment]
@anwar6953 a few more (minor ?) comments. However, please get the guice changes reviewed with @chtyim. May be able to do better there? Also, because of the guice changes, it would be good to run integration tests before merging.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java (hunk 1 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java:120]: Hmm, I think I bound it in DataSets module, because that's where the bindings are configured for LineageWriter in terms of inmemory, standalone, vs distributed.
This override is a subset of distributed (for program runnables).

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/AbstractRemoteSystemOpsHandler.java (hunk 7 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/AbstractRemoteSystemOpsHandler.java:64]: I'm not sure if there's a difference in this case.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java (hunk 8 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteLineageWriterHandler.java:50]: will do it in a separate PR to keep this PR smaller.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteOpsClient.java (hunk 10 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/remote/RemoteOpsClient.java:97]: doesn't need to be static, but I think @Nishith had suggested it in previous PR.
It makes sense to indicate that the method doesn't rely on nor affect state.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java (hunk 13 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java:129]: What do you mean if its ok to "refactor them as..." ?
Whether that will work? Or whether certain bindings are unnecessary?

File: cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java (hunk 12 above)
Comment (anwar6953) [inline cdap-explore/src/main/java/co/cask/cdap/hive/context/ContextManager.java]: Ah, I need to validate whether the bindings there are appropriate.
Will do that.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java (hunk 13 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java:129]: right, seems like there are some extra bindings.

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java (hunk 1 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/DistributedProgramRunnableModule.java:120]: this looks a little strange to me, but I don't have a better suggestion, other than splitting the `DataSetsModule`. At the very least, a `DataSetsModules#getProgramRunnerModules()` instead of doing overriding like this - although I don't know if that is less strange. Perhaps @chtyim can advice.

**Comment by anwar6953** [comment]
Resolved conflicts and rerunning build (link in PR description).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java (hunk 13 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/DistributedMapReduceTaskContextProvider.java:129]: I checked these. Some have just been bundled into modules.
For instance, there was previously the explicit binding of a Logging class. Now, its using the LoggingModule#getDistributedModules, which has the same binding in it.
There are a few unnecessary bindings, but based upon testing, they don't seem to hurt (the code is much cleaner this way).

**Comment by anwar6953** [comment]
Build passed - Merging
http://builds.cask.co/browse/CDAP-DUT4255-10
